src/firefly/js/core/MasterSaga.js

```diff
@@ -16,18 +21,99 @@ export function dispatchAddSaga(saga, params={}) {
+ * @param {string} id  a unique identifier of the watcher to cancel
+ */
+export function dispatchCancelWatcher(id) {
+    flux.process({ type: CANCEL_WATCHER, payload: {id}});
+}
+
+/**
+ * Cancel all watchers.  Should only be called during re-init scenarios. 
+ */
+export function dispatchCancelAllWatchers() {
+    flux.process({ type: CANCEL_ALL_WATCHER});
+}
+
 
 /**
  * This saga launches all the predefined Sagas then loops and waits for any ADD_SAGA actions and launches those Segas
  */
 export function* masterSaga() {
-
+    let watchers = {};
 
     // Start a saga from any action
     while (true) {
-        var action= yield take(ADD_SAGA);
-        const {getState, dispatch}= flux.getRedux();
-        const {saga,params}= action.payload;
-        if (typeof saga === 'function') yield fork( saga, params, dispatch, getState);
+        const action= yield take([ADD_SAGA, ADD_WATCHER, CANCEL_WATCHER, CANCEL_ALL_WATCHER]);
+
+        switch (action.type) {
+            case ADD_SAGA: {
+                const {getState, dispatch}= flux.getRedux();
+                const {saga,params}= action.payload;
+                if (typeof saga === 'function') yield fork( saga, params, dispatch, getState);
+                break;
+            }
+            case ADD_WATCHER: {
+                const {getState, dispatch}= flux.getRedux();
+                const {id, actions, callback}= action.payload;
+                if (id && actions && callback) {
+                    if (watchers[id]) {
+                        yield cancel(watchers[id]);
+                    }
+                    const saga = createSaga({id, actions, callback});
+                    const task = yield fork(saga, dispatch, getState);
+                    watchers[id] = task;
+                    isDebug() && console.log(`watcher ${id} added.  #watcher: ${Object.keys(watchers).length}`);
+                }
+                break;
+            }
+            case CANCEL_WATCHER: {
+                const {id}= action.payload;
+                const task = watchers[id];
+                if (task) {
+                    yield cancel(task);
+                    Reflect.deleteProperty(watchers, id);
+                    isDebug() && console.log(`watcher ${id} cancelled.  #watcher: ${Object.keys(watchers).length}`);
+                }
+                break;
+            }
+            case CANCEL_ALL_WATCHER: {
+                const ids = Object.keys(watchers);
+                for (let i = 0; i < ids.length; i++) {
+                    const task = watchers[ids[i]];
+                    yield cancel(task);
+                }
+                watchers = {};
+                break;
+            }
+        }
     }
-}
\ No newline at end of file
+}
+
+
+function createSaga({id, actions=[], callback}) {
+    const saga = function* () {
+        while (true) {
+            const action= yield take(actions);
+            try {
+                callback && callback(action);
```

Comment:
This watcher could be simplify and mostly replace the way we use sagas if we called the callback with following parametes:

1. action
1. cancelSelf
1. params
1. dispatch
1. getState

i.e.

`callback(action, ()=> dispatch({ type: CANCEL_WATCHER, payload: {id}}), params, dispatch, getState)`